iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/command_line_parser.hpp

```diff
@@ -0,0 +1,235 @@
+
+  private:
+    struct Argument
+    {
+        char shortId;
+        name_t id;
+        value_t value;
+    };
+
+    binaryName_t m_binaryName;
+    cxx::vector<Argument, MAX_NUMBER_OF_ARGUMENTS> m_arguments;
+};
+
+/// @brief Factory class for the CommandLineOptions. First one has to register
+///        all switches and options before calling parse. This is required for
+///        the help page which is generated and printed on failure as well as
+///        for consistency and syntax checks.
+class CommandLineParser
+{
+  public:
+    static constexpr uint64_t MAX_DESCRIPTION_LENGTH = 1024;
+    static constexpr uint64_t OPTION_OUTPUT_WIDTH = 45;
+    static constexpr uint64_t MAX_TYPE_LENGTH = 16;
+    static constexpr char NO_SHORT_OPTION = '\0';
+
+    using description_t = cxx::string<MAX_DESCRIPTION_LENGTH>;
+    using typeName_t = cxx::string<MAX_TYPE_LENGTH>;
+
+    /// @brief The constructor.
+    /// @param[in] programDescription The description to the program. Will be printed in the help.
+    /// @param[in] onFailureCallback callback which is called when parse fails, if nothing is
+    ///            defined std::exit(EXIT_FAILURE) is called
+    explicit CommandLineParser(
+        const description_t& programDescription,
+        const cxx::function<void()> onFailureCallback = [] { std::exit(EXIT_FAILURE); }) noexcept;
+
+    /// @brief Adds a command line switch argument
+    ///        Calls the error handler when the option was already added or the shortOption and longOption are empty.
+    /// @param[in] shortOption a single letter as short option
+    /// @param[in] longOption a multi letter word which does not start with minus as long option name
+    /// @param[in] description the description to the argument
+    CommandLineParser& addSwitch(const char shortOption,
+                                 const CommandLineOptions::name_t& longOption,
+                                 const description_t& description) noexcept;
+
+    /// @brief Adds a command line optional value argument.
+    ///        Calls the error handler when the option was already added or the shortOption and longOption are empty.
+    /// @param[in] shortOption a single letter as short option
+    /// @param[in] longOption a multi letter word which does not start with minus as long option name
+    /// @param[in] description the description to the argument
+    /// @param[in] typeName the name of the value type
+    /// @param[in] defaultValue the value which will be set to the option when it is not set by the user
+    CommandLineParser& addOptionalValue(const char shortOption,
+                                        const CommandLineOptions::name_t& longOption,
+                                        const description_t& description,
+                                        const typeName_t& typeName,
+                                        const CommandLineOptions::value_t& defaultValue) noexcept;
```

Comment:
This is only partially correct. This API is an internal API and the user will never use it, they will use the public macro which supports actually what you are proposing.
If I would implement this my nice forward technique to `addOption` would break - therefore I propose to leave it as it is.